IBAN code -> stacktrace

Try to speak DE12345678901234567890 and MaryTTS will fail with a stacktrace.

This is critical because IBAN codes are often used.

The stacktrace is with version 5.1 but you can reproduce it in the demo (http://mary.dfki.de:59125/), too.

```
<html><body><h1>Internal server error</h1><p>Processing failed.<pre>java.lang.Exception: Module Preprocess: Problem processing the data.
        at marytts.server.Request.processOneChunk(Request.java:551)
        at marytts.server.Request.processOrLookupOneChunk(Request.java:389)
        at marytts.server.Request.process(Request.java:336)
        at marytts.server.http.SynthesisRequestHandler.process(SynthesisRequestHandler.java:272)
        at marytts.server.http.SynthesisRequestHandler.handleClientRequest(SynthesisRequestHandler.java:98)
        at marytts.server.http.BaseHttpRequestHandler.handle(BaseHttpRequestHandler.java:144)
        at org.apache.http.nio.protocol.BufferingHttpServiceHandler$RequestHandlerAdaptor.handle(BufferingHttpServiceHandler.java:189)
        at org.apache.http.nio.protocol.SimpleNHttpRequestHandler.handle(SimpleNHttpRequestHandler.java:51)
        at org.apache.http.nio.protocol.AsyncNHttpServiceHandler.processRequest(AsyncNHttpServiceHandler.java:453)
        at org.apache.http.nio.protocol.AsyncNHttpServiceHandler.requestReceived(AsyncNHttpServiceHandler.java:225)
        at org.apache.http.nio.protocol.BufferingHttpServiceHandler.requestReceived(BufferingHttpServiceHandler.java:127)
        at org.apache.http.impl.nio.DefaultNHttpServerConnection.consumeInput(DefaultNHttpServerConnection.java:161)
        at org.apache.http.impl.nio.DefaultServerIOEventDispatch.inputReady(DefaultServerIOEventDispatch.java:147)
        at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:161)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:335)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:275)
        at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
        at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:542)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NumberFormatException: For input string: "12345678901234567890"
        at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
        at java.lang.Long.parseLong(Long.java:444)
        at java.lang.Long.valueOf(Long.java:513)
        at java.lang.Long.decode(Long.java:665)
        at marytts.language.de.preprocess.NumberEP.expandInteger(NumberEP.java:279)
        at marytts.language.de.preprocess.AbbrevEP.expandAbbrev(AbbrevEP.java:186)
        at marytts.language.de.preprocess.AbbrevEP.expandAbbrev(AbbrevEP.java:203)
        at marytts.language.de.preprocess.AbbrevEP.expand(AbbrevEP.java:127)
        at marytts.language.de.preprocess.ExpansionPattern.process(ExpansionPattern.java:275)
        at marytts.language.de.Preprocess.matchAndExpandPatterns(Preprocess.java:118)
        at marytts.language.de.Preprocess.process(Preprocess.java:69)
        at marytts.server.Request.processOneChunk(Request.java:549)
        ... 19 more
</pre></body></html>

--- a/marytts-languages/marytts-lang-de/src/main/java/marytts/language/de/preprocess/AbbrevEP.java
+++ b/marytts-languages/marytts-lang-de/src/main/java/marytts/language/de/preprocess/AbbrevEP.java
@@ -184,8 +184,14 @@ public class AbbrevEP extends ExpansionPattern {
 			String text = MaryDomUtils.tokenText(token);
 			// Only digits? Pronounce as an integer.
 			if (REPattern.onlyDigits.matcher(text).find()) {
-				logger.debug("Expanding as integer: `" + text + "'");
-				exp.addAll(makeNewTokens(token.getOwnerDocument(), number.expandInteger(text)));
+				if (Pattern.matches(NumberEP.sInteger, text)) {
+					logger.debug("Expanding as integer: `" + text + "'");
+					exp.addAll(makeNewTokens(token.getOwnerDocument(), number.expandInteger(text)));
+				} else {
+					logger.debug("Expanding as digits: `" + text + "'");
+					exp.addAll(makeNewTokens(token.getOwnerDocument(), number.expandDigits(text)));
+				}
+
 			} else if (text.length() > 1) {
 				logger.debug("Expanding one token by rule: `" + text + "'");
 				// Slow down the mtu containing this token (or the token

--- a/marytts-languages/marytts-lang-de/src/test/java/marytts/language/de/PreprocessorIT.java
+++ b/marytts-languages/marytts-lang-de/src/test/java/marytts/language/de/PreprocessorIT.java
@@ -71,4 +71,19 @@ public class PreprocessorIT extends MaryModuleTestCase {
 		processAndCompare("unicode1", Locale.GERMAN);
 	}
 
+	@Test
+	public void testIBAN1() throws Exception {
+		processAndCompare("iban1", Locale.GERMAN);
+	}
+
+	@Test
+	public void testIBAN2() throws Exception {
+		processAndCompare("iban2", Locale.GERMAN);
+	}
+
+	@Test(expected=NumberFormatException.class)
+	public void testIBAN3() throws Exception {
+		processAndCompare("iban3", Locale.GERMAN);
+	}
+
 }

--- a/marytts-languages/marytts-lang-de/src/test/java/marytts/language/de/PreprocessorIT.java
+++ b/marytts-languages/marytts-lang-de/src/test/java/marytts/language/de/PreprocessorIT.java
@@ -81,7 +81,7 @@ public class PreprocessorIT extends MaryModuleTestCase {
 		processAndCompare("iban2", Locale.GERMAN);
 	}
 
-	@Test(expected=Exception.class)
+	@Test
 	public void testIBAN3() throws Exception {
 		processAndCompare("iban3", Locale.GERMAN);
 	}

--- a/marytts-languages/marytts-lang-de/src/test/java/marytts/language/de/PreprocessorIT.java
+++ b/marytts-languages/marytts-lang-de/src/test/java/marytts/language/de/PreprocessorIT.java
@@ -81,7 +81,7 @@ public class PreprocessorIT extends MaryModuleTestCase {
 		processAndCompare("iban2", Locale.GERMAN);
 	}
 
-	@Test(expected=NumberFormatException.class)
+	@Test(expected=Exception.class)
 	public void testIBAN3() throws Exception {
 		processAndCompare("iban3", Locale.GERMAN);
 	}

--- a/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban1.preprocessed
+++ b/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban1.preprocessed
@@ -1,0 +1,127 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<maryxml xmlns="http://mary.dfki.de/2002/MaryXML" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="0.5" xml:lang="de">
+<p>
+<s>
+<mtu orig="GR16 0110 1250 0000 0001 2300 695">
+<mtu accent="first" orig="GR16">
+<prosody rate="-20%">
+<phonology precision="precise">
+<mtu accent="last" orig="GR">
+<t accent="unknown">
+G
+</t>
+<t accent="unknown">
+R
+</t>
+</mtu>
+</phonology>
+</prosody>
+<t>
+sechzehn
+</t>
+</mtu>
+<mtu accent="last" orig="01">
+<t>
+null
+</t>
+<t accent="unknown">
+eins
+</t>
+</mtu>
+<boundary breakindex="3" tone="unknown"/>
+<mtu accent="last" orig="10">
+<t>
+eins
+</t>
+<t accent="unknown">
+null
+</t>
+</mtu>
+<boundary breakindex="3" tone="unknown"/>
+<mtu accent="last" orig="12">
+<t>
+eins
+</t>
+<t accent="unknown">
+zwei
+</t>
+</mtu>
+<boundary breakindex="3" tone="unknown"/>
+<mtu accent="last" orig="50">
+<t>
+fünf
+</t>
+<t accent="unknown">
+null
+</t>
+</mtu>
+<boundary breakindex="3" tone="unknown"/>
+<mtu accent="last" orig="00">
+<t>
+null
+</t>
+<t accent="unknown">
+null
+</t>
+</mtu>
+<boundary breakindex="3" tone="unknown"/>
+<mtu accent="last" orig="00">
+<t>
+null
+</t>
+<t accent="unknown">
+null
+</t>
+</mtu>
+<boundary breakindex="3" tone="unknown"/>
+<mtu accent="last" orig="00">
+<t>
+null
+</t>
+<t accent="unknown">
+null
+</t>
+</mtu>
+<boundary breakindex="3" tone="unknown"/>
+<mtu accent="last" orig="01">
+<t>
+null
+</t>
+<t accent="unknown">
+eins
+</t>
+</mtu>
+<boundary breakindex="3" tone="unknown"/>
+<mtu accent="last" orig="23">
+<t>
+zwei
+</t>
+<t accent="unknown">
+drei
+</t>
+</mtu>
+<boundary breakindex="3" tone="unknown"/>
+<mtu accent="last" orig="00">
+<t>
+null
+</t>
+<t accent="unknown">
+null
+</t>
+</mtu>
+<boundary breakindex="3" tone="unknown"/>
+<mtu accent="last" orig="695">
+<t>
+sechs
+</t>
+<t>
+neun
+</t>
+<t accent="unknown">
+fünf
+</t>
+</mtu>
+</mtu>
+</s>
+</p>
+</maryxml>

--- a/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban1.tokenised
+++ b/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban1.tokenised
@@ -1,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<maryxml xmlns="http://mary.dfki.de/2002/MaryXML" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="0.5" xml:lang="de">
+<p>
+<s>
+<mtu orig="GR16 0110 1250 0000 0001 2300 695">
+<t>
+GR16
+</t>
+<t>
+0110
+</t>
+<t>
+1250
+</t>
+<t>
+0000
+</t>
+<t>
+0001
+</t>
+<t>
+2300
+</t>
+<t>
+695
+</t>
+</mtu>
+</s>
+</p>
+</maryxml>

--- a/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban2.preprocessed
+++ b/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban2.preprocessed
@@ -1,0 +1,86 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<maryxml xmlns="http://mary.dfki.de/2002/MaryXML" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="0.5" xml:lang="de">
+<p>
+<s>
+<mtu orig="GB29 NWBK 6016 1331 9268 19">
+<mtu accent="first" orig="GB29">
+<prosody rate="-20%">
+<phonology precision="precise">
+<mtu accent="last" orig="GB">
+<t accent="unknown">
+G
+</t>
+<t accent="unknown">
+B
+</t>
+</mtu>
+</phonology>
+</prosody>
+<t>
+neunundzwanzig
+</t>
+</mtu>
+<prosody rate="-20%">
+<phonology precision="precise">
+<mtu accent="last" orig="NWBK">
+<t accent="unknown">
+N
+</t>
+<t accent="unknown">
+W
+</t>
+<t accent="unknown">
+B
+</t>
+<t accent="unknown">
+K
+</t>
+</mtu>
+</phonology>
+</prosody>
+<t>
+sechs
+</t>
+<t>
+Tausend
+</t>
+<t>
+sechzehn
+</t>
+<t>
+ein
+</t>
+<t>
+Tausend
+</t>
+<t>
+drei
+</t>
+<t>
+Hundert
+</t>
+<t>
+einunddreißig
+</t>
+<t>
+neun
+</t>
+<t>
+Tausend
+</t>
+<t>
+zwei
+</t>
+<t>
+Hundert
+</t>
+<t>
+achtundsechzig
+</t>
+<t>
+neunzehn
+</t>
+</mtu>
+</s>
+</p>
+</maryxml>

--- a/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban2.tokenised
+++ b/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban2.tokenised
@@ -1,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<maryxml xmlns="http://mary.dfki.de/2002/MaryXML" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="0.5" xml:lang="de">
+<p>
+<s>
+<mtu orig="GB29 NWBK 6016 1331 9268 19">
+<t>
+GB29
+</t>
+<t>
+NWBK
+</t>
+<t>
+6016
+</t>
+<t>
+1331
+</t>
+<t>
+9268
+</t>
+<t>
+19
+</t>
+</mtu>
+</s>
+</p>
+</maryxml>

--- a/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban3.preprocessed
+++ b/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban3.preprocessed
@@ -1,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<maryxml xmlns="http://mary.dfki.de/2002/MaryXML" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="0.5" xml:lang="de">
+<p>
+<s>
+<t>
+DE12345678901234567890
+</t>
+</s>
+</p>
+</maryxml>

--- a/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban3.preprocessed
+++ b/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban3.preprocessed
@@ -2,9 +2,80 @@
 <maryxml xmlns="http://mary.dfki.de/2002/MaryXML" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="0.5" xml:lang="de">
 <p>
 <s>
-<t>
-DE12345678901234567890
+<mtu accent="first" orig="DE12345678901234567890">
+<prosody rate="-20%">
+<phonology precision="precise">
+<mtu accent="last" orig="DE">
+<t accent="unknown">
+D
 </t>
+<t accent="unknown">
+E
+</t>
+</mtu>
+</phonology>
+</prosody>
+<t>
+eins
+</t>
+<t>
+zwei
+</t>
+<t>
+drei
+</t>
+<t>
+vier
+</t>
+<t>
+fünf
+</t>
+<t>
+sechs
+</t>
+<t>
+sieben
+</t>
+<t>
+acht
+</t>
+<t>
+neun
+</t>
+<t>
+null
+</t>
+<t>
+eins
+</t>
+<t>
+zwei
+</t>
+<t>
+drei
+</t>
+<t>
+vier
+</t>
+<t>
+fünf
+</t>
+<t>
+sechs
+</t>
+<t>
+sieben
+</t>
+<t>
+acht
+</t>
+<t>
+neun
+</t>
+<t>
+null
+</t>
+</mtu>
 </s>
 </p>
 </maryxml>

--- a/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban3.tokenised
+++ b/marytts-languages/marytts-lang-de/src/test/resources/marytts/language/de/iban3.tokenised
@@ -1,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<maryxml xmlns="http://mary.dfki.de/2002/MaryXML" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="0.5" xml:lang="de">
+<p>
+<s>
+<t>
+DE12345678901234567890
+</t>
+</s>
+</p>
+</maryxml>

